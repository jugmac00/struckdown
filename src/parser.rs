use std::ops::Range;

use lazy_static::lazy_static;
use pulldown_cmark as cm;
use regex::Regex;

use crate::event::{Alignment, Attrs, Event, EventType, Location, Str, Tag};

lazy_static! {
    static ref TEXT_ROLE_RE: Regex = Regex::new(r"\{([^\r\n\}]+)\}$").unwrap();
    static ref CODE_ROLE_RE: Regex = Regex::new(r"^\{([^\r\n\}]+)\}(?:\s+(.*?))$").unwrap();
    static ref HEADING_ID_RE: Regex = Regex::new(r"{#([^\r\n\}]+)\}\s*$").unwrap();
}

/// Reads until the end of an image tag all embedded content as raw string.
fn read_raw_alt<'a, 'b, I: Iterator<Item = (cm::Event<'b>, Range<usize>)>>(
    iter: &'a mut I,
) -> String {
    let mut depth = 1;
    let mut rv = String::new();
    while let Some((event, _)) = iter.next() {
        match event {
            cm::Event::Start(..) => depth += 1,
            cm::Event::End(..) => depth -= 1,
            cm::Event::Text(text) => rv.push_str(&text),
            cm::Event::Code(code) => rv.push_str(&code),
            cm::Event::SoftBreak | cm::Event::HardBreak => rv.push('\n'),
            _ => {}
        }
        if depth == 0 {
            break;
        }
    }
    rv
}

/// Parses a structured cmark document into an event stream.
pub fn parse(s: &str) -> impl Iterator<Item = Event> {
    let mut opts = cm::Options::empty();
    opts.insert(cm::Options::ENABLE_TABLES);
    opts.insert(cm::Options::ENABLE_STRIKETHROUGH);
    opts.insert(cm::Options::ENABLE_TASKLISTS);
    opts.insert(cm::Options::ENABLE_FOOTNOTES);

    let parser = cm::Parser::new_with_broken_link_callback(s, opts, None);
    let mut iter = parser.into_offset_iter().peekable();
    let mut tag_stack = vec![];
    let mut pending_role = None;

    std::iter::from_fn(move || {
        if let Some((event, range)) = iter.next() {
            // inefficient way to find the location
            let mut loc = Location {
                offset: range.start,
                len: range.end - range.start,
                line: s[..range.start].chars().filter(|&c| c == '\n').count() + 1,
                column: match s[..range.start].rfind('\n') {
                    Some(nl) => range.start - nl - 1,
                    None => range.start,
                },
            };

            // simple events
            let ty = match event {
                cm::Event::Start(cm_tag) => {
                    let mut attrs = Attrs::default();
                    let tag = match cm_tag {
                        cm::Tag::Paragraph => Tag::Paragraph,
                        cm::Tag::Heading(1) => Tag::Heading1,
                        cm::Tag::Heading(2) => Tag::Heading2,
                        cm::Tag::Heading(3) => Tag::Heading3,
                        cm::Tag::Heading(4) => Tag::Heading4,
                        cm::Tag::Heading(5) => Tag::Heading5,
                        cm::Tag::Heading(6) => Tag::Heading6,
                        cm::Tag::Heading(_) => unreachable!(),
                        cm::Tag::BlockQuote => Tag::BlockQuote,
                        cm::Tag::CodeBlock(kind) => match kind {
                            cm::CodeBlockKind::Fenced(lang) => {
                                let lang: Str<'_> = lang.into();
                                if let Some(m) = CODE_ROLE_RE.captures(lang.as_str()) {
                                    let g1 = m.get(1).unwrap();
                                    let g2 = m.get(2).unwrap();
                                    attrs.role = Some(lang.slice(g1.start(), g1.end()));
                                    attrs.argument = Some(lang.slice(g2.start(), g2.end()));
                                    Tag::Directive
                                } else {
                                    attrs.argument = Some(lang);
                                    Tag::FencedCode
                                }
                            }
                            cm::CodeBlockKind::Indented => Tag::IndentedCode,
                        },
                        cm::Tag::List(None) => Tag::UnorderedList,
                        cm::Tag::List(Some(start)) => {
                            attrs.start = Some(start as u32);
                            Tag::OrderedList
                        }
                        cm::Tag::Item => Tag::ListItem,
                        cm::Tag::FootnoteDefinition(id) => {
                            attrs.id = Some(id.into());
                            Tag::FootnoteDefinition
                        }
                        cm::Tag::Table(alignments) => {
                            attrs.alignments = Some(
                                alignments
                                    .into_iter()
                                    .map(|cm_align| match cm_align {
                                        cm::Alignment::None => Alignment::None,
                                        cm::Alignment::Left => Alignment::Left,
                                        cm::Alignment::Center => Alignment::Center,
                                        cm::Alignment::Right => Alignment::Right,
                                    })
                                    .collect(),
                            );
                            Tag::Table
                        }
                        cm::Tag::TableHead => Tag::TableHead,
                        cm::Tag::TableRow => Tag::TableRow,
                        cm::Tag::TableCell => Tag::TableCell,
                        cm::Tag::Emphasis => Tag::Emphasis,
                        cm::Tag::Strong => Tag::Strong,
                        cm::Tag::Strikethrough => Tag::Strikethrough,
                        cm::Tag::Link(_, target, title) => {
                            attrs.target = Some(target.into());
                            if !title.is_empty() {
                                attrs.title = Some(title.into());
                            }
                            Tag::Link
                        }
                        cm::Tag::Image(_, target, title) => {
                            // images are special in that we downgrade them from
                            // tags to toplevel events to not have to deal with
                            // nested text.
                            let alt = read_raw_alt(&mut iter);
                            return Some(Event::new(
                                EventType::Image {
                                    target: target.into(),
                                    alt: if alt.is_empty() {
                                        None
                                    } else {
                                        Some(alt.into())
                                    },
                                    title: if title.is_empty() {
                                        None
                                    } else {
                                        Some(title.into())
                                    },
                                },
                                loc,
                            ));
                        }
                    };
                    tag_stack.push(tag);
                    EventType::StartTag { tag, attrs }
                }
                cm::Event::End(_) => EventType::EndTag {
                    tag: tag_stack.pop().unwrap(),
                },
                cm::Event::Text(text) => {
                    let mut text: Str<'_> = text.into();

                    // handle roles
                    if let Some(&(cm::Event::Code(_), _)) = iter.peek() {
                        if let Some(m) = TEXT_ROLE_RE.captures(text.as_str()) {
                            let g0 = m.get(0).unwrap();
                            let g1 = m.get(1).unwrap();

                            // adjust the span of the text to not include the role.
                            let column_adjustment = g0.end() - g0.start();
                            loc.len -= column_adjustment;
                            pending_role =
                                Some((text.slice(g1.start(), g1.end()), column_adjustment));
                            text = text.slice(0, g0.start());
                        }
                    }
                    EventType::Text { text }
                }
                cm::Event::Code(value) => {
                    // if there is a pending role then we're not working with a
                    // code block, but an interpreted text one.
                    if let Some((role, column_adjustment)) = pending_role.take() {
                        loc.column -= column_adjustment;
                        loc.offset -= column_adjustment;
                        loc.len += column_adjustment;
                        EventType::InterpretedText {
                            text: value.into(),
                            role: role.into(),
                        }
                    } else {
                        EventType::InlineCode { code: value.into() }
                    }
                }
                cm::Event::Html(html) => EventType::RawHtml { html: html.into() },
                cm::Event::FootnoteReference(target) => EventType::FootnoteReference {
                    target: target.into(),
                },
                cm::Event::SoftBreak => EventType::SoftBreak,
                cm::Event::HardBreak => EventType::HardBreak,
                cm::Event::Rule => EventType::Rule,
                cm::Event::TaskListMarker(checked) => EventType::Checkbox { checked },
            };

            Some(Event::new(ty, loc))
        } else {
            None
        }
    })
}
