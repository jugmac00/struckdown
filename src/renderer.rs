use std::collections::HashMap;
use std::io::{self, Write};

use v_htmlescape::escape;

use crate::event::{
    Alignment, AnnotatedEvent, Attrs, CheckboxEvent, CodeBlockEvent, DirectiveEvent, EndTagEvent,
    Event, FootnoteReferenceEvent, ImageEvent, InlineCodeEvent, InterpretedTextEvent, RawHtmlEvent,
    StartTagEvent, Str, Tag, TextEvent,
};

pub struct HtmlRenderer<'data, F> {
    out: F,
    footnotes: HashMap<Str<'data>, usize>,
}

fn is_block_tag(tag: Tag) -> bool {
    match tag {
        Tag::Paragraph => true,
        Tag::Heading1 => true,
        Tag::Heading2 => true,
        Tag::Heading3 => true,
        Tag::Heading4 => true,
        Tag::Heading5 => true,
        Tag::Heading6 => true,
        Tag::BlockQuote => true,
        Tag::OrderedList => true,
        Tag::UnorderedList => true,
        Tag::ListItem => true,
        Tag::FootnoteDefinition => true,
        Tag::Table => true,
        Tag::TableHead => true,
        Tag::TableRow => true,
        Tag::TableCell => true,
        Tag::Emphasis => false,
        Tag::Strong => false,
        Tag::Strikethrough => false,
        Tag::Link => false,
        Tag::TableHeader => true,
        Tag::TableBody => true,
    }
}

impl<'data, F: Write> HtmlRenderer<'data, F> {
    pub fn new(out: F) -> HtmlRenderer<'data, F> {
        HtmlRenderer {
            out,
            footnotes: HashMap::new(),
        }
    }

    fn tag_to_html_tag(&self, tag: Tag) -> &'static str {
        match tag {
            Tag::Paragraph => "p",
            Tag::Heading1 => "h1",
            Tag::Heading2 => "h2",
            Tag::Heading3 => "h3",
            Tag::Heading4 => "h4",
            Tag::Heading5 => "h5",
            Tag::Heading6 => "h6",
            Tag::BlockQuote => "blockquote",
            Tag::OrderedList => "ol",
            Tag::UnorderedList => "ul",
            Tag::ListItem => "li",
            Tag::FootnoteDefinition => "div",
            Tag::Table => "table",
            Tag::TableHeader => "thead",
            Tag::TableBody => "tbody",
            Tag::TableRow => "tr",
            Tag::TableHead => "th",
            Tag::TableCell => "td",
            Tag::Emphasis => "em",
            Tag::Strong => "strong",
            Tag::Strikethrough => "ss",
            Tag::Link => "a",
        }
    }

    fn start_tag(&mut self, tag: Tag, attrs: &Attrs) -> Result<(), io::Error> {
        let html_tag = self.tag_to_html_tag(tag);
        write!(self.out, "<{}", html_tag)?;

        match attrs.start {
            None | Some(1) => {}
            Some(start) => {
                write!(self.out, " start={}", start)?;
            }
        }

        let mut combined_style = String::new();
        if let Some(ref id) = attrs.id {
            write!(self.out, " id=\"{}\"", escape(id.as_str()))?;
        }
        if let Some(ref title) = attrs.title {
            write!(self.out, " title=\"{}\"", escape(title.as_str()))?;
        }
        if let Some(ref target) = attrs.target {
            write!(self.out, " href=\"{}\"", escape(target.as_str()))?;
        }

        combined_style.push_str(match attrs.alignment {
            Alignment::None => "",
            Alignment::Left => "text-align: left",
            Alignment::Center => "text-align: center",
            Alignment::Right => "text-align: right",
        });

        if let Some(ref custom) = attrs.custom {
            for (key, value) in custom.iter() {
                if key == "style" {
                    if !combined_style.is_empty() {
                        combined_style.push_str("; ");
                    }
                    combined_style.push_str(value.as_str());
                } else {
                    write!(self.out, " {}=\"{}\"", key, escape(value.as_str()))?;
                }
            }
        }

        if !combined_style.is_empty() {
            write!(self.out, " style=\"{}\"", escape(&combined_style))?;
        }

        write!(self.out, ">")?;

        Ok(())
    }

    fn end_tag(&mut self, tag: Tag) -> Result<(), io::Error> {
        let html_tag = self.tag_to_html_tag(tag);

        write!(
            self.out,
            "</{}>{}",
            html_tag,
            if is_block_tag(tag) { "\n" } else { "" }
        )?;

        Ok(())
    }

    pub fn event(&mut self, event: &AnnotatedEvent<'data>) -> Result<(), io::Error> {
        match event.event {
            Event::StartTag(StartTagEvent { tag, ref attrs }) => {
                self.start_tag(tag, attrs)?;
            }
            Event::EndTag(EndTagEvent { tag }) => {
                self.end_tag(tag)?;
            }
            Event::Text(TextEvent { ref text }) => {
                write!(self.out, "{}", escape(text.as_str()))?;
            }
            Event::CodeBlock(CodeBlockEvent {
                ref code,
                ref language,
            }) => {
                write!(self.out, "<pre><code")?;
                if let Some(language) = language {
                    write!(self.out, " class=\"lang-{}\"", language.as_str())?;
                }
                write!(self.out, ">{}</code></pre>", escape(code.as_str()))?;
            }
            Event::Directive(DirectiveEvent {
                ref name, ref body, ..
            }) => {
                write!(
                    self.out,
                    "<div class=\"directive-{}\"><pre>{}</pre></div>",
                    escape(name.as_str()),
                    escape(body.as_str()),
                )?;
            }
            Event::InterpretedText(InterpretedTextEvent { ref text, ref role }) => {
                write!(
                    self.out,
                    "<span class=\"role-{}\">{}</span>",
                    escape(role.as_str()),
                    escape(text.as_str())
                )?;
            }
            Event::InlineCode(InlineCodeEvent { ref code }) => {
                write!(self.out, "<code>{}</code>", escape(code.as_str()))?;
            }
            Event::Image(ImageEvent {
                ref target,
                ref alt,
                ref title,
            }) => {
                write!(
                    self.out,
                    "<img src=\"{}\" alt=\"{}\" title=\"{}\">",
                    target,
                    alt.as_ref().map_or("", |x| x.as_str()),
                    title.as_ref().map_or("", |x| x.as_str()),
                )?;
            }
            Event::RawHtml(RawHtmlEvent { ref html }) => {
                write!(self.out, "{}", html)?;
            }
            Event::SoftBreak => {}
            Event::HardBreak => {}
            Event::Rule => {
                write!(self.out, "<hr>")?;
            }
            Event::Checkbox(CheckboxEvent { checked }) => {
                write!(
                    self.out,
                    "<input type=checkbox disabled{}>",
                    if checked { " checked" } else { "" }
                )?;
            }
            Event::FootnoteReference(FootnoteReferenceEvent { ref target }) => {
                let number = match self.footnotes.get(target) {
                    Some(&num) => num,
                    None => {
                        let next_number = self.footnotes.len() + 1;
                        self.footnotes.insert(target.clone(), next_number);
                        next_number
                    }
                };
                write!(
                    self.out,
                    "<sup class=footnote-reference><a href=\"#{}\">{}</a></sup>",
                    escape(target.as_str()),
                    number,
                )?;
            }
        }
        Ok(())
    }
}

pub fn to_html<'a, I: Iterator<Item = AnnotatedEvent<'a>>>(iter: I) -> String {
    let mut out = Vec::<u8>::new();
    let mut renderer = HtmlRenderer::new(&mut out);
    for event in iter {
        renderer.event(&event).unwrap();
    }
    unsafe { String::from_utf8_unchecked(out) }
}
