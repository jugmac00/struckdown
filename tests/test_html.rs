use std::fs;

use structured_cmark::parser::parse;
use structured_cmark::renderer::to_html;

#[test]
fn test_basics() {
    insta::glob!("inputs/*.md", |file| {
        let source = fs::read_to_string(file).unwrap();
        let html = to_html(parse(&source));
        insta::assert_snapshot!(html);
    });
}