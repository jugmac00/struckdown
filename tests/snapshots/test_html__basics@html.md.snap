---
source: tests/test_html.rs
expression: html
input_file: tests/inputs/html.md
---
<p>This document has some inline html:</p>
<div class=foo>
  bar
</div>
<p>And this has some mixed markdown:</p>
<div class=foo>
<p><em>markdown here</em></p>
</div>
